brave/src/main/java/brave/Tracer.java

```diff
@@ -44,7 +44,35 @@
  * @see Span
  * @see Propagation
  */
-public final class Tracer {
+public final class Tracer implements Closeable {
+  // volatile for visibility on get. writes guarded by Tracer.class
+  static volatile Tracer current = null;
+
+  /**
+   * Returns the most recent tracer that hasn't been closed.
+   *
+   * <p>This object should not be cached.
+   */
+  @Nullable public static Tracer current() {
```

Comment:
Here's what the code looks like for p6spy (and mysql which are the only instrumentation so far I have seen need this)

```java
@Override public void onBeforeAnyExecute(StatementInformation info) {
    Tracer tracer = Tracer.current();
    if (tracer == null) return;

    Span span = tracer.nextSpan();
    ...
```

One option is to create a noop tracer, or some composite that allows none. This affects the codebase more for what it a special case (statically initialized trace instrumentation). While not objective, I think it also encourages a pattern we should discourage, which is to rely on `Tracer.current()` for routine work. If we did, we'd need to rethink what that would return, or make a delegate type.

Ex with a special lazy initialized tracer (which implies defining what empty means)
```java
   // note tracer currently has luxury of being a final type.
   // This implies we make something that doles out span ids even in noop case!
   // Doing so could break propagation when a tracer is eventually initialized.
   final Tracer currentTracer = Tracer.current(); // can be static or instance

@Override public void onBeforeAnyExecute(StatementInformation info) {
    Span span = tracer.nextSpan();
    ...
```

If we did a callback approach, I think it would be a bit more complex, especially from a GC perspective, as we'd have a registration (ex. MySQL would register with an intermediary for tracer update callbacks). I think the callsite of the instrumentation would be slightly less cluttered, but the machinery would be substantial and perhaps maintenance causing.

```java
// static or instance, but needs to be volatile since callback can happen on any thread.
volatile Tracer currentTracer; 

@Override public void onCurrentTracer(Tracer currentTracer) {
   this.currentTracer = currentTracer;
}

@Override public void onBeforeAnyExecute(StatementInformation info) {
    if (currentTracer == null) return;

    Span span = tracer.nextSpan();
    ...
```

With the proposed, we can make our instrumentation indirect access to Tracer.current() anyway, availing users to manage how they see fit. Users could use something framework specific that might be a better fit (ex static injection).

```java
/** Override to use a mechanism besides Tracer.current(), such as static injection */
@Nullable protected Tracer currentTracer() {
  return Tracer.current();
}

@Override public void onBeforeAnyExecute(StatementInformation info) {
    Tracer tracer = currentTracer();
    if (tracer == null) return;

    Span span = tracer.nextSpan();
    ...
```

wdyt?